### PR TITLE
feat: add miner mode change by method

### PR DESCRIPTION
**Pull request by gabriel-aranha-cw**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Implemented thread-safe miner mode changes using `RwLock<MinerMode>` in the Miner struct.
- Added new RPC method `stratus_change_miner_mode` for changing miner mode.
- Introduced new error variants to handle various miner mode change scenarios.
- Added interval miner shutdown management in the `GlobalState`.
- Implemented comprehensive integration tests for miner mode changes.
- Updated E2E workflow to include miner tests.
- Refactored executor and miner configuration to use the new thread-safe miner mode.
- Added helper function for leader-follower synchronization in E2E tests.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>9 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>executor.rs</strong><dd><code>Refactor miner lock acquisition in executor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/executor/executor.rs

<li>Modified the miner lock acquisition logic to use a read lock on the <br>miner mode<br> <li> Adjusted the conditional logic for acquiring the miner lock based on <br>the new read lock<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-6b460d7dee8280c0287e8d9d9d59cf66a57ec9fed1bc003c6ede6fef60c7376b">+16/-8</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>miner.rs</strong><dd><code>Implement thread-safe miner mode with RwLock</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/miner/miner.rs

<li>Changed <code>mode</code> field to use <code>RwLock<MinerMode></code> instead of <code>MinerMode</code><br> <li> Updated <code>new</code> method to initialize <code>mode</code> with <code>RwLock</code><br> <li> Modified <code>spawn_interval_miner</code> to use read lock on <code>mode</code><br> <li> Removed <code>mode()</code> method<br> <li> Updated <code>save_execution</code> to use read lock on <code>mode</code><br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-16f448afc7dae3e8e802f676a86dbd7051e19f6c17cbbeb53eb730d726372629">+43/-18</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>miner_config.rs</strong><dd><code>Update MinerMode and init_miner for thread-safety</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/miner/miner_config.rs

<li>Updated <code>MinerMode</code> derive attributes to include <code>PartialEq</code>, <br><code>serde::Deserialize</code><br> <li> Modified <code>init_miner</code> to use read lock on <code>mode</code><br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-ea3a8597b909e696993f7e791ab04e5ad784da8077b87088d9c78c804e1ef69d">+12/-3</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>stratus_error.rs</strong><dd><code>Add new error variants for miner mode changes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/primitives/stratus_error.rs

<li>Added new error variants: <code>RpcTransactionEnabled</code>, <br><code>PendingTransactionsExist</code>, <code>MinerModeConflict</code>, <code>MinerModeLockFailed</code>, <br><code>MinerModeChangeUnsupported</code>, <code>MinerModeParamInvalid</code>, <code>ConsensusSet</code><br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-301cded01d772388e3e5c08ede093b91c3e4478c08e11a48f70e3d44351385fb">+31/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rpc_server.rs</strong><dd><code>Add RPC method for changing miner mode</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/rpc/rpc_server.rs

<li>Added <code>stratus_change_miner_mode</code> RPC method<br> <li> Updated error handling for consensus and miner mode locks<br> <li> Implemented logic for changing miner mode<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-835ba255c9a5c1fe0e13285bc39e058e1c74422e5e03ebba483c9d8a15c45405">+113/-4</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>globals.rs</strong><dd><code>Add interval miner shutdown management</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/globals.rs

<li>Added <code>INTERVAL_MINER_SHUTDOWN</code> static variable<br> <li> Implemented methods for managing interval miner shutdown state<br> <li> Updated <code>GlobalState</code> struct and related methods to include interval <br>miner shutdown state<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-48d6fbc3a4ed67100952dcccfada858623c931e73d6de32984d4d1457e68d3a9">+34/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rpc.ts</strong><dd><code>Add helper function for leader-follower synchronization</code>&nbsp; &nbsp; </dd></summary>
<hr>

e2e/cloudwalk-contracts/integration/test/helpers/rpc.ts

<li>Added <code>waitForFollowerToSyncWithLeader</code> function to synchronize leader <br>and follower nodes<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-bfbd0a045ec4243ba68015b78cb7fb7cb485ff47e44bc3c22e3e155fc3ddf003">+20/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>leader-follower-importer.test.ts</strong><dd><code>Refactor leader-follower importer test</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

e2e/cloudwalk-contracts/integration/test/leader-follower-importer.test.ts

<li>Removed <code>waitForFollowerToSyncWithLeader</code> function (moved to helpers)<br> <li> Updated imports to use the new helper function<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-76e205db03980cadded8f3f9ff7be76d1f3d3af18d4d8bf2cc8fa09d74934b8d">+6/-19</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>e2e-leader-follower.yml</strong><dd><code>Include miner tests in E2E workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/e2e-leader-follower.yml

- Added 'miner' to the test matrix

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-6cb08ce033aeec24672fd91944789df5188b47bbf56f16d034c600ba66ea1461">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>leader-follower-miner.test.ts</strong><dd><code>Add integration tests for miner mode changes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

e2e/cloudwalk-contracts/integration/test/leader-follower-miner.test.ts

<li>Added new integration test file for miner mode changes<br> <li> Implemented tests for various miner mode change scenarios<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1675/files#diff-4b1c5fc9f33c8088178dffe2234e70df03667bf90fa3a5663fd17c3dabfe13a5">+170/-0</a>&nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: src/eth/rpc/rpc_server.rs [hunk 1]
```diff
@@ -258,7 +261,12 @@ async fn stratus_health(_: Params<'_>, context: Arc<RpcContext>, _: Extensions)
     let should_serve = match GlobalState::get_node_mode() {
         NodeMode::Leader => true,
         NodeMode::Follower => {
-            let consensus_lock = context.consensus.read().map_err(|_| StratusError::ConsensusLockFailed)?;
+            let consensus_lock = context.consensus.read().map_err(|poison| {
+                tracing::error!("consensus read lock was poisoned");
+                context.consensus.clear_poison();
+                drop(poison.into_inner());
```
Comment (marcospb19-cw) [inline src/eth/rpc/rpc_server.rs:267]: nitpick: These drops are redundant as a local variable always drops by the end of the function (before the function returns any value)

Comment (gabriel-aranha-cw) [reply src/eth/rpc/rpc_server.rs:267]: I receive warning if I do not use drop.
```
unused `std::sync::RwLockReadGuard` that must be used
if unused the RwLock will immediately unlock
`#[warn(unused_must_use)]` on by defaultrustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B0%5D?0#file%3A%2F%2F%2FUsers%2Fgabrielmartins%2FCloudWalk%2Fstratus%2Fsrc%2Feth%2Fminer%2Fminer.rs)
miner.rs(92, 17): use `let _ = ...` to ignore the resulting value: `let _ = `
```

I figured using drop is cleaner than using the `let _ = ...`

Comment (gabriel-aranha-cw) [reply src/eth/rpc/rpc_server.rs:267]: Is something like this equivalent?
```
let consensus_lock = context.consensus.read().map_err(|_| {
                tracing::error!("consensus read lock was poisoned");
                context.consensus.clear_poison();
```

**Pull request merged**